File: MatlabCode/06_b/DP_IC.m
%% Dynamic programming for inventory control (stochastic)

close all;
clear all;
clc;



%% Parameters and initialization

% Time horizon
%    slides: t = 0,...,T-1
%      code: t = 1,...,T
T = 100;

% Time horizon
TT=[1:T]'; 

% Load some parameters defined in file DP_IC_setup.m:
%   max capcity (C), demand prob (pW), stage cost coeff. (cost), 
%   max demand (W_max) and possible demand values (W_set)
[pW,cost,C,W_max,W_set]=DP_IC_setup();

% Number of states = C+1
N_state = C+1;

% Initializize some matrices to store results

% Matrix to store optimal policy 
% U(x,t) optimal input if in state x at time t
U=zeros(N_state,T);

% Optimal value functions (cost-to-go functions)
% V(x,t) cost-to-go if in state x at time t
V = zeros(N_state, T+1);



%% Compute Optimal Policy

[U, V] = DP_IC_optimal_policy(T);



%% Simulate Optimal Policy (1 run)

% Initial state: full
x0=C;

% Single run simulation
[x, u, gt, w] = DP_IC_singlerun(T,U,x0);



%% Display results

% Write some results
disp(' ');
disp(['Time horizon T = ' num2str(T) ]);
disp(' ');
disp(['Total cost = ' num2str(sum(gt))]); 
disp(['Expected total cost from x0 (value function V0(x0)) = ' num2str(sum(V(end,1)))]);
disp(['Average stage cost = ' num2str(mean(gt))]);
disp(['First reorder at time instant ' num2str(find(u,1)-1)]);
disp(['Average time between consecutive reorders ' num2str(mean(diff(find(u)))) ]);

% Plot x vs. time
hold on;
stairs([TT; T+1],x);
xlabel('time');
ylabel('x');
title('State (stock level)');
grid;


% Plot u vs. time
figure
hold on;
stairs(TT,u);
xlabel('time');
ylabel('u');
title('Input (ordered quantity)');
grid;

% Plot gt vs. time
figure
hold on;
bar(TT,gt);
xlabel('time');
ylabel('gt');
title('Stage cost');
xlim([0 T]);
grid;

% Plot J vs. time
figure
hold on;
area(TT,cumsum(gt),'FaceColor','r');
xlabel('time');
ylabel('J');
title('Total cost');
grid;

File: PythonCode/06_b/DP_IC.py
from DP_IC_utils import *


# parameters and initialization
# time horizon
T = 100
# load parameters
pW, cost, C, W_max, W_set = DP_IC_setup()

# compute optimal policy (1 run)
U, V = DP_IC_optimal_policy(T)

# simulate optimal policy (1 run)
x0 = C      # initially full
x, u, gt, w = DP_IC_singlerun(T, U, x0)

# display results
print(f'Time horizon T: {T}')
print(f'Total cost = {np.sum(gt)}')
print(f'Expected total cost from x0 (value function V0(x0)) = {np.sum(V[-1, 1])}')
print(f'Average stage cost = {np.mean(gt)}')
print(f'First reorder time: {next((i for i, x in enumerate(u) if x), None)}')
print(f'Average reordering time: {np.mean(np.diff(np.array([i for i, x in enumerate(u) if x])))}')
print(f'Total number of orders: {np.count_nonzero(u)}')

# plot
fig, axs = plt.subplots(2, 2)
fig.suptitle('Simulation')

axs[0, 0].stairs(x, range(T + 2))
axs[0, 0].set_title('State (stock level)')

axs[0, 1].bar(range(T), u)
axs[0, 1].set_title('Input (ordered quantity)')

axs[1, 0].bar(range(T), gt)
axs[1, 0].set_title('Stage cost')

axs[1, 1].fill_between(range(T), np.cumsum(gt))
axs[1, 1].set_title('Total cost')

plt.show()
